From 54b2b1eb4ef798215cccebc281e3425021e6984b Mon Sep 17 00:00:00 2001
From: Chris Rogers <rogersc@ainfosec.com>
Date: Wed, 26 Oct 2016 12:06:54 -0400
Subject: [PATCH 3/6] libxl-crypto-key-dir.patch

* Support passing a crypto key directory for a guest's disk(s). Otherwise uses
  platform default path


Signed-off-by: Chris Rogers <rogersc@ainfosec.com>
---
 tools/libxl/libxl.c          | 2 +-
 tools/libxl/libxl_blktap2.c  | 8 +++++++-
 tools/libxl/libxl_dm.c       | 2 +-
 tools/libxl/libxl_internal.h | 3 ++-
 tools/libxl/libxl_types.idl  | 3 +++
 tools/libxl/xl_cmdimpl.c     | 3 +++
 6 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/tools/libxl/libxl.c b/tools/libxl/libxl.c
index f8ef5af..79e8965 100644
--- a/tools/libxl/libxl.c
+++ b/tools/libxl/libxl.c
@@ -2564,7 +2564,7 @@ static void device_disk_add(libxl__egc *egc, uint32_t domid,
             case LIBXL_DISK_BACKEND_TAP:
                 if (dev == NULL) {
                     dev = libxl__blktap_devpath(gc, disk->pdev_path,
-                                                disk->format);
+                                                disk->format, d_config.b_info.crypto_key_dir);
                     if (!dev) {
                         LOG(ERROR, "failed to get blktap devpath for %p",
                             disk->pdev_path);
diff --git a/tools/libxl/libxl_blktap2.c b/tools/libxl/libxl_blktap2.c
index 83138df..5219948 100644
--- a/tools/libxl/libxl_blktap2.c
+++ b/tools/libxl/libxl_blktap2.c
@@ -25,7 +25,8 @@ int libxl__blktap_enabled(libxl__gc *gc)
 
 char *libxl__blktap_devpath(libxl__gc *gc,
                             const char *disk,
-                            libxl_disk_format format)
+                            libxl_disk_format format,
+							char *keydir)
 {
     const char *type;
     char *params, *devname = NULL;
@@ -40,6 +41,11 @@ char *libxl__blktap_devpath(libxl__gc *gc,
             return devname;
     }
 
+	if(!strcmp(keydir, ""))
+	    setenv("TAPDISK2_CRYPTO_KEYDIR", "/config/platform-crypto-keys", 1);
+	else
+		setenv("TAPDISK2_CRYPTO_KEYDIR", keydir, 1);
+
     params = libxl__sprintf(gc, "%s:%s", type, disk);
     err = tap_ctl_create(params, &devname);
     if (!err) {
diff --git a/tools/libxl/libxl_dm.c b/tools/libxl/libxl_dm.c
index 2d5eea4..8179093 100644
--- a/tools/libxl/libxl_dm.c
+++ b/tools/libxl/libxl_dm.c
@@ -1142,7 +1142,7 @@ static int libxl__build_device_model_args_new(libxl__gc *gc,
                 if (disks[i].backend == LIBXL_DISK_BACKEND_TAP) {
                     format = qemu_disk_format_string(LIBXL_DISK_FORMAT_RAW);
                     pdev_path = libxl__blktap_devpath(gc, disks[i].pdev_path,
-                                                      disks[i].format);
+                                                      disks[i].format, b_info->crypto_key_dir);
                 } else {
                     pdev_path = disks[i].pdev_path;
                 }
diff --git a/tools/libxl/libxl_internal.h b/tools/libxl/libxl_internal.h
index 8acb913..0274177 100644
--- a/tools/libxl/libxl_internal.h
+++ b/tools/libxl/libxl_internal.h
@@ -1685,7 +1685,8 @@ _hidden int libxl__blktap_enabled(libxl__gc *gc);
  */
 _hidden char *libxl__blktap_devpath(libxl__gc *gc,
                                     const char *disk,
-                                    libxl_disk_format format);
+                                    libxl_disk_format format,
+                                    char *keydir);
 
 /* libxl__device_destroy_tapdisk:
  *   Destroys any tapdisk process associated with the backend represented
diff --git a/tools/libxl/libxl_types.idl b/tools/libxl/libxl_types.idl
index 8d96fa7..547af2e 100644
--- a/tools/libxl/libxl_types.idl
+++ b/tools/libxl/libxl_types.idl
@@ -432,6 +432,9 @@ libxl_domain_build_info = Struct("domain_build_info",[
     ("cpuid",           libxl_cpuid_policy_list),
     ("blkdev_start",    string),
 
+	#directory containing the crypto keys for the vm's disks
+	("crypto_key_dir", string),
+
     ("vnuma_nodes", Array(libxl_vnode_info, "num_vnuma_nodes")),
     
     ("device_model_version", libxl_device_model_version),
diff --git a/tools/libxl/xl_cmdimpl.c b/tools/libxl/xl_cmdimpl.c
index c35b2c4..10c14b3 100644
--- a/tools/libxl/xl_cmdimpl.c
+++ b/tools/libxl/xl_cmdimpl.c
@@ -1480,6 +1480,9 @@ static void parse_config_data(const char *config_source,
     if (!xlu_cfg_get_long(config, "max_event_channels", &l, 0))
         b_info->event_channels = l;
 
+	if (!xlu_cfg_get_string (config, "crypto_key_dir", &buf, 0))
+		xlu_cfg_replace_string(config, "crypto_key_dir", &b_info->crypto_key_dir, 0);
+
     xlu_cfg_replace_string (config, "kernel", &b_info->kernel, 0);
     xlu_cfg_replace_string (config, "ramdisk", &b_info->ramdisk, 0);
     xlu_cfg_replace_string (config, "device_tree", &b_info->device_tree, 0);
-- 
2.1.4

